File: CPP/ByTier/D/2137_D_Replace_with_Occurrences.cpp
// Problem: CF 2137 D - Replace with Occurrences
// https://codeforces.com/contest/2137/problem/D

/*
D. Replace with Occurrences

Purpose:
This code determines whether there exists an array a of size n such that for each element b[i], 
the count of occurrences of a[i] in the array equals b[i]. If such an array exists, it constructs one.
The solution uses grouping by frequency and assigning unique identifiers to groups.

Algorithms/Techniques:
- Hash map to group indices by their values
- Greedy assignment of identifiers based on grouping by frequency

Time Complexity:
O(n) average case, where n is the size of input array. Each element is processed once and hash map operations are amortized O(1).

Space Complexity:
O(n) for storing the array elements and auxiliary structures (hash map and result vector).
*/

#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <utility>
#include <vector>

using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(0);
  int T;
  cin >> T;
  while (T--) {
    int x;
    cin >> x;
    vector<int> y(x);
    for (int i = 0; i < x; i++)
      cin >> y[i];
    unordered_map<int, vector<int>> z;
    for (int i = 0; i < x; i++)
      z[y[i]].push_back(i);
    vector<int> w(x, -1);
    int p = 1, q = 1;
    for (auto u = z.begin(); u != z.end(); ++u) {
      int v = u->first;
      auto &r = u->second;
      if ((int)r.size() % v) { // Check if the count of occurrences is divisible by the value itself
        q = 0;
        break;
      }
      for (int i = 0; i < (int)r.size(); i += v) { // Assign identifiers in batches
        for (int j = 0; j < v; j++)
          w[r[i + j]] = p; // Assign same identifier to group of indices
        p++;
      }
    }
    if (!q)
      cout << "-1\n";
    else {
      for (int i = 0; i < x; i++)
        cout << w[i] << " ";
      cout << "\n";
    }
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/D/2137_D_Replace_with_Occurrences.py
# Problem: CF 2137 D - Replace with Occurrences
# https://codeforces.com/contest/2137/problem/D

"""
D. Replace with Occurrences

Purpose:
This code determines whether there exists an array 'a' such that the frequency of each element in 'a'
matches the given array 'b'. If such an array exists, it constructs one; otherwise, it outputs -1.

Algorithm/Techniques:
- Greedy approach with frequency tracking
- Uses a mapping from elements to their assigned values based on occurrence counts
- Tracks how many times each number can still be used before being assigned a new value

Time Complexity: O(n) - Each element is processed once.
Space Complexity: O(n) - For storing arrays and counts.
"""

import sys

input = sys.stdin.readline
minput = lambda: map(int, input().split())

for _ in range(int(input())):
    n = int(input())
    cnt = [0] * (n + 1)          # Tracks how many times each number can be reused
    ans = [0] * n                # Output array to be constructed
    a = 0                        # Counter for assigning new values
    num = [0] * (n + 1)          # Maps numbers to their assigned values
    
    b = list(minput())
    
    for i in range(n):
        if cnt[b[i]] == 0:
            # First time seeing this number, assign a new value
            cnt[b[i]] = b[i] - 1   # Set how many more times we can reuse this value
            a += 1                 # Increment counter for new assignment
            num[b[i]] = a          # Assign the new value to this number
        else:
            # Reducing the remaining count before reusing
            cnt[b[i]] -= 1
        
        ans[i] = num[b[i]]       # Assign current value to answer array
    
    if sum(cnt) == 0:
        print(*ans)
    else:
        print(-1)


# https://github.com/VaHiX/codeForces/